examples/vars.yml

```diff
@@ -0,0 +1,25 @@
+postgres_password: "{{ lookup('password', 'inventory/host_vars/{{domain}}/passwords/postgres.psk chars=ascii_letters,digits') }}"
+
+# You can set any pict-rs environmental variables here. They will populate the templates/docker-compose.yml file.
+# https://git.asonix.dog/asonix/pict-rs
+pictrs_env_vars:
+  - PICTRS__SERVER__API_KEY: "{{ postgres_password }}"
+  - PICTRS__MEDIA__VIDEO_CODEC: vp9
+  - PICTRS__MEDIA__GIF__MAX_WIDTH: 256
+  - PICTRS__MEDIA__GIF__MAX_HEIGHT: 256
+  - PICTRS__MEDIA__GIF__MAX_AREA: 65536
+  - PICTRS__MEDIA__GIF__MAX_FRAME_COUNT: 400
+  - PICTRS_OPENTELEMETRY_URL: http://otel:4137
+  - RUST_LOG: debug
+  - RUST_BACKTRACE: full
+#  - PICTRS__STORE__TYPE: object_storage 
+#  - PICTRS__STORE__ENDPOINT: '<S3 endpoint>'
+#  - PICTRS__STORE__BUCKET_NAME: '<bucket name>'
+#  - PICTRS__STORE__REGION: '<region>'
+#  - PICTRS__STORE__USE_PATH_STYLE: false 
+#  - PICTRS__STORE__ACCESS_KEY: '<access key>'
+#  - PICTRS__STORE__SECRET_KEY: '<secret key>'
+
+
```

Comment:
How about using a .env file instead? Then you can include the vars in docker-compose.yml with a single line `env_file`

Reply:
If we go that route, we would still want Ansible to generate the `env_file`(s) accordingly. It allows people to modify and configure the installation in a single centralized place (my goal is to clean this up and get everything in the same spot, then document everything so users have more control over their installations). I think this is a better UX and more "Ansible" oriented.

If this gets merged, I'd do the same thing for `lemmy-ui` immediately and plop all available configuration variables in `vars.yml`, which also resolves #44.  

What are your thoughts on populating the respective `.env` file(s) via Ansible and updating the template to point to the respective `.env` file(s) we generate?

Reply:
I think the `yml` file is fine, ansible seems to prefer those anyway.

Reply:
Yes makes sense to move all env vars into a separate file, so that people dont have to edit docker-compose.yml. It also makes it easier to secure passwords by restricting permissions of this file. I would definitely prefer using .env as its much easier (only a single line needed to import).

Reply:
The permissions aren't a big concern as the way the playbooks are structured is they're run locally--meaning any sensitive variables stay local to your system. Making a `.env` file `0600` or a `vars.yml` file `0600` is trivial. 

We could use a singular generated .env file for all VMs, but it's a little ugly as it'd populate all the environmental variables for all VMs, which could lead to unexpected results. It also doesn't allow for much flexibility--for example, if I want to set `RUST_LOG` to a different level on two Docker VMs, it wouldn’t be possible with your proposal. 

To do this properly, each VM would have its own environment file. A user must copy the sample environment files to the appropriate directory, edit it, and then run the playbook in contrast to this solution, where a user needs to modify a single file while Ansible does the rest. 

Let me see if I can come up with something that satisfies what you’re looking for without compromising UX.

Reply:
What are your thoughts on this approach? This will allow the best mix of UX and flexibility for instance operators. 

In the `docker-compose.yml` file, we would reference the respective `.env` files (they would be put in `{{ lemmy_base_dir }}/inventory/host_vars/{{ domain }}/lemmyui.env,pictrs.env,postgres.env,etc`. 

This change also allows operators to create/modify docker environment variables in the future without us having to alter the playbooks each time or have to distribute and maintain `.env` sample templates. 

#### Example playbook

The variables references would be located in `vars.yml`. This is just an example of the proposed logic. 

```yaml
---
- hosts: localhost
  become: no
  vars:
    postgres_password: IAteTooManySourPatchKids
    docker_env_vars:
      - name: "pictrs"
        variables:
          PICTRS__SERVER__API_KEY: "{{ postgres_password }}"
          PICTRS__MEDIA__VIDEO_CODEC: vp9
          PICTRS__MEDIA__GIF__MAX_WIDTH: 256
          PICTRS__MEDIA__GIF__MAX_HEIGHT: 256
          PICTRS__MEDIA__GIF__MAX_AREA: 65536
          PICTRS__MEDIA__GIF__MAX_FRAME_COUNT: 400
          PICTRS_OPENTELEMETRY_URL: http://otel:4137
          RUST_LOG: debug
          RUST_BACKTRACE: full
      #   PICTRS__STORE__TYPE: object_storage 
      #   PICTRS__STORE__ENDPOINT: '<S3 endpoint>'
      #   PICTRS__STORE__BUCKET_NAME: '<bucket name>'
      #   PICTRS__STORE__REGION: '<region>'
      #   PICTRS__STORE__USE_PATH_STYLE: false 
      #   PICTRS__STORE__ACCESS_KEY: '<access key>'
      #   PICTRS__STORE__SECRET_KEY: '<secret key>'
      - name: "lemmyui"
        variables:
          LEMMY_UI_HTTPS: true
      - name: "postgres"
        variables:
          POSTGRES_USER: lemmy
          POSTGRES_PASSWORD: "{{ postgres_password }}"
          POSTGRES_DB: lemmy
          
  tasks:
    - name: Create environment variable files
      copy:
        dest: "{{ item.name }}.env"
        content: |
          {% for key, value in item.variables.items() %}
          {{ key }}={{ value }}
          {% endfor %}
      loop: "{{ docker_env_vars }}"
```

```bash
<snip>
TASK [Create environment variable files] ************************************************************
changed: [localhost] => (item={'name': 'pictrs', 'variables': {'PICTRS__SERVER__API_KEY': 'IAteTooManySourPatchKids', 'PICTRS__MEDIA__VIDEO_CODEC': 'vp9', 'PICTRS__MEDIA__GIF__MAX_WIDTH': 256, 'PICTRS__MEDIA__GIF__MAX_HEIGHT': 256, 'PICTRS__MEDIA__GIF__MAX_AREA': 65536, 'PICTRS__MEDIA__GIF__MAX_FRAME_COUNT': 400, 'PICTRS_OPENTELEMETRY_URL': 'http://otel:4137', 'RUST_LOG': 'debug', 'RUST_BACKTRACE': 'full'}})
changed: [localhost] => (item={'name': 'lemmyui', 'variables': {'LEMMY_UI_HTTPS': True}})
changed: [localhost] => (item={'name': 'postgres', 'variables': {'POSTGRES_USER': 'lemmy', 'POSTGRES_PASSWORD': 'IAteTooManySourPatchKids', 'POSTGRES_DB': 'lemmy'}})
</snip>
```

```shell
bash-3.2$ ls *.env
lemmyui.env     pictrs.env      postgres.env

bash-3.2$ cat lemmyui.env
LEMMY_UI_HTTPS=True

bash-3.2$ cat pictrs.env
PICTRS__SERVER__API_KEY=IAteTooManySourPatchKids
PICTRS__MEDIA__VIDEO_CODEC=vp9
PICTRS__MEDIA__GIF__MAX_WIDTH=256
PICTRS__MEDIA__GIF__MAX_HEIGHT=256
PICTRS__MEDIA__GIF__MAX_AREA=65536
PICTRS__MEDIA__GIF__MAX_FRAME_COUNT=400
PICTRS_OPENTELEMETRY_URL=http://otel:4137
RUST_LOG=debug
RUST_BACKTRACE=full

bash-3.2$ cat postgres.env
POSTGRES_USER=lemmy
POSTGRES_PASSWORD=IAteTooManySourPatchKids
POSTGRES_DB=lemmy
```

If you and @dessalines  prefer this method, I'll go ahead and get everything done and merged in :).

Reply:
I made this change in another branch/PR in my repo so you can see what it would look like. It's ready to merge and has been tested: https://github.com/codyro/lemmy-ansible/pull/3/files

Reply:
I personally prefer this PR to your new one that creates additional `.env` files. 

Mainly because the latter *duplicates* settings, so it might not be clear that when you edit one of those `.env` files, and run the playbook, it will actually just regenerate them from the `vars.yml` file. So they are really useless files IMO.

Reply:
I'm in agreement, although I don't have strong feelings either way. 

I'll let you and @Nutomic decide the approach you guys would prefer to merge.

I'm also open to other ideas if anyone has any!

Reply:
Having a separate .env file for each service seems totally unnecessary. All the ones I see already have prefixes PICTRS, POSTGRES etc so there can be no confusion. In case of RUST_LOG, you can define LEMMY_RUST_LOG in .env and then rename it like this:
```
environment:
  RUST_LOG: ${LEMMY_RUST_LOG}
```

Reply:
IMO the less places we have to configure, the better, so I'd much prefer this single `vars.yml` you've done here.

Reply:
> I'd much prefer this single `vars.yml`

@codyro I agree. Singular variable file per "host" is the simplest approach that will cause the least amount of issues.

Having them inside the playbook is a non-starter because the whole point of moving the variables out is so that users of the playbooks do not have to edit playbooks. The goal is for them to make changes outside the playbooks (env_file) and then run the playbook.

The lack of defaults is the only thing that annoys me. ie; what happens if a user were to forget this crucial step of copying the variables.